[ALLUXIO-2552] Support symbolic permissions in chmod command

rvesse opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-2552

This provides the initial support for specifying permissions in symbolic form when using the `chmod` command. This allows for specifying them more explicitly without users having to know how to convert them into the octal numeric form e.g.

    > alluxio fs chmod u=rw,go=r /example

Syntax is based upon the standard *nix symbolic syntax with some limitations, namely:

- The permissions `X`, `s`, and `t` are not supported since they make no sense in Alluxio
- Permissions cannot be specified in mutative form i.e. `g+w` or `u-x`

Supporting mutative permissions would require more substantial changes because permissions are always applied declaratively currently. The associated JIRA contains a description of a possible future enhancement to support this.

The functionality for parsing the symbolic form is provided in its own class so it can also be used by users if desired.  The `chmod` command is updated to use this to parse in the desired permissions. Unit test tests are provided as well as integration tests. I may still add some further test cases for the new functionality to this request.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13342/
Test PASSed.

ifcharming commented on core/common/src/test/java/alluxio/security/authorization/ModeParserTest.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  @Test
+  public void symbolics_separated() {
```
nit: s/symbolics_separated/symbolicsSeparated

ifcharming commented on core/common/src/test/java/alluxio/security/authorization/ModeParserTest.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  @Test
+  public void symbolics_combined() {
```
nit: s/symbolics_combined/symbolicsCombined

ifcharming commented on core/common/src/test/java/alluxio/security/authorization/ModeParserTest.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  @Test
+  public void symbolics_cumulative() {
```
nit: s/symbolics_cumulative/symbolicsCumulative

ifcharming commented on core/common/src/test/java/alluxio/security/authorization/ModeParserTest.java:
```diff
@@ -0,0 +1,110 @@
+ * @author rvesse
+ *
+ */
+public final class ModeParserTest {
```
can you please add another test case for partial chmod, such as
`u=rwx` 
`go=rw`

rvesse replied:
Done

rvesse commented:
I have addressed the comments and added a number of additional positive and negative test cases.

I am waiting on legal approval before I sign the contributor agreement. I am happy to sign it personally and I have managerial approval to sign, I just need our legal department to review it first.

I have made contributions in the past prior to the current process and I have signed a similar agreement with the Apache software foundation so I don't expect there to be any issues in that regard, it mainly depends how long the lawyers take to review it.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13349/
Test PASSed.

ifcharming commented:
LGTM

Thank you @rvesse for the enhancement! Once the CLA is signed, I'll approve and merge this PR.

apc999 commented on core/common/src/main/java/alluxio/security/authorization/ModeParser.java:
```diff
@@ -0,0 +1,127 @@
+ * @author rvesse
+ *
+ */
+public final class ModeParser {
```
NIT: `@ThreadSafe` annotation

rvesse replied:
Done

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13355/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13356/
Test PASSed.

ifcharming commented:
LGTM
